source/sigv4.c

```diff
@@ -1243,57 +1237,59 @@ static SigV4Status_t generateCredentialScope( const SigV4Parameters_t * pSigV4Pa
         assert( canonicalURILen != NULL );
         assert( *canonicalURILen > 0U );
 
-        pUriLoc = pUri;
-        pBufLoc = pCanonicalURI;
         bufferLen = *canonicalURILen;
 
-        while( ( i++ < uriLen ) && ( returnStatus == SigV4Success ) )
+        for( ; ( uriIndex < uriLen ) && ( returnStatus == SigV4Success ); uriIndex++ )
         {
-            if( doubleEncodeEquals && ( *pUriLoc == '=' ) )
+            if( doubleEncodeEquals && ( pUri[ uriIndex ] == '=' ) )
             {
                 if( ( bufferLen - bytesConsumed ) < URI_DOUBLE_ENCODED_EQUALS_CHAR_SIZE )
                 {
                     returnStatus = SigV4InsufficientMemory;
-                    LOG_INSUFFICIENT_MEMORY_ERROR( "encode the URI",
+                    LOG_INSUFFICIENT_MEMORY_ERROR( "double encode '=' character in canonical query",
                                                    ( bytesConsumed + URI_DOUBLE_ENCODED_EQUALS_CHAR_SIZE - bufferLen ) );
                 }
                 else
                 {
-                    bytesConsumed += writeDoubleEncodedEquals( pBufLoc, bufferLen - bytesConsumed );
+                    bytesConsumed += writeDoubleEncodedEquals( pCanonicalURI + bytesConsumed, bufferLen - bytesConsumed );
                 }
             }
-            else if( isAllowedChar( *pUriLoc, encodeSlash ) )
+            else if( isAllowedChar( pUri[ uriIndex ], encodeSlash ) )
             {
+                /* If the output buffer has space, add the character as-is in URI encoding as it
+                 * is neither a special character nor an '=' character requiring double encoding. */
                 if( bytesConsumed < bufferLen )
                 {
-                    *pBufLoc = *pUriLoc;
-                    ++pBufLoc;
+                    pCanonicalURI[ bytesConsumed ] = pUri[ uriIndex ];
                     ++bytesConsumed;
                 }
                 else
                 {
                     returnStatus = SigV4InsufficientMemory;
-                    LOG_INSUFFICIENT_MEMORY_ERROR( "encode the URI", bytesConsumed + 1U - bufferLen );
+                    LogError( ( "Failed to encode URI in buffer due to insufficient memory" ) );
```

Comment:
Why remove the insufficient memory error macro? It's there to log this exact situation